lib/engine/game/g_1880/step/check_fi_connection.rb

```diff
@@ -0,0 +1,128 @@
+
+          def actions(entity)
+            return [] unless entity == current_entity
+            return [] unless entity.minor?
+            return ['choose'] if @merging
+
+            ACTIONS
+          end
+
+          def description
+            'Foreign investor merge'
+          end
+
+          def blocks?
+            true
+          end
+
+          def auto_actions(entity)
+            return [] if @merging
+            return [Engine::Action::Pass.new(entity)] unless @game.check_for_foreign_investor_connection(entity)
+
+            [Engine::Action::DestinationConnection.new(entity)]
+          end
+
+          def process_destination_connection(action)
+            corporation = action.entity.shares.first.corporation
+            @merging = [corporation, action.entity]
+            @game.log << "#{action.entity.full_name} and #{corporation.name} are connected"
+            process_merger
+
+            if action.entity.cash.positive?
+              @state = :choose_precent
+            else
+              process_precent(Action::Choose.new(action.entity,
+                                                 choice: '100%'))
+            end
+          end
+
+          def skip!
+            pass!
+          end
+
+          def process_merger
+            fi = @merging[1]
+            share = @merging[1].shares.first
+            @game.share_pool.transfer_shares(share.to_bundle, fi.owner)
+            @game.bank.spend(50, fi.owner)
+            @game.log << "#{fi.owner.name} recieves #{@game.format_currency(50)} and a share of #{share.corporation.name}"
+          end
+
+          def process_choose(action)
+            case @state
+            when :choose_precent
+              process_precent(action)
+            when :choose_token
+              process_token(action)
+            end
+          end
+
+          def setup
+            @state = nil
+            @merging = nil
+          end
+
+          def process_precent(action)
+            fi = @merging[1]
+            if fi.cash.positive?
+              amount = action.choice == '100%' ? fi.cash : (fi.cash * 20) / 100
+              fi.spend(amount, @merging[0], check_positive: false)
+              @game.log << "#{fi.full_name} transfers #{@game.format_currency(amount)} to #{@merging[0].name}"
+            end
+            process_token(Action::Choose.new(fi, choice: 'Discard')) unless cheapest_unused_token(@merging[0])
+            @state = :choose_token
+          end
+
+          def process_token(action)
+            fi = @merging[1]
+            replace_token(fi, fi.tokens.first, cheapest_unused_token(@merging[0])) if action.choice == 'Replace'
+            fi.tokens.first.city.remove_tokens! if action.choice == 'Discard'
+            @game.log << "#{fi.full_name} closes"
+            fi.close!
+
+            pass!
+          end
+
+          def cheapest_unused_token(corp)
+            corp.tokens.find { |t| t.city.nil? }
```

Comment:
This method isn't deterministic. It's relying on tokens being in cost order, which is true today and probably tomorrow, but not guaranteed.

Reply:
corp.tokens.find(&:unused).min_by(&:cost)